### ISSUE TITLE
3proxy падает сразу после запуска

### ISSUE THREAD

**Issue by srvh** [3proxy/3proxy#722]
3proxy падает сразу после старта службы на CentOS 7
journalctl -xe показывает такие строки:

kernel: 3proxy[23628]: segfault at 0 ip 0000000000408b32 sp 00002b0228b26120 error 4
systemd[1]: 3proxy.service: main process exited, code=killed, status=11/SEGV
systemd[1]: Unit 3proxy.service entered failed state.
systemd[1]: 3proxy.service failed.

kernel: 3proxy[23365]: segfault at 0 ip 00002ad16049ae71 sp 00002ad161a521e8 error 4 in libc-2.17.so[2ad16040e000+1c4000]
systemd[1]: 3proxy.service: main process exited, code=killed, status=11/SEGV
systemd[1]: Unit 3proxy.service entered failed state.
systemd[1]: 3proxy.service failed.

*** Error in `/bin/3proxy': double free or corruption (fasttop): 0x00002abb7c0009b0 ***
systemd[1]: 3proxy.service: main process exited, code=killed, status=6/ABRT
systemd[1]: Unit 3proxy.service entered failed state.
systemd[1]: 3proxy.service failed.

Пробовал ставить в конфиге 
stacksize 1024000

Не помогает. В чем может быть проблема?

**Comment by srvh** [comment]
Чаще всего вот эта ошибка в логах
3proxy[16519]: *** Error in `/bin/3proxy': double free or corruption (fasttop): 0x00007f9ef003a380 ***

**Comment by z3APA3A** [comment]
Какая версия 3proxy? Какой конфиг? Какие плагины? Если 3proxy апдейтился и используются плагины - обновлялись ли они после обновления 3proxy?

**Comment by srvh** [comment]
Чаще всего вот эта строка появляется.
*** Error in `/bin/3proxy': double free or corruption (fasttop): 0x0000000002584080 ***
3proxy.service: main process exited, code=killed, status=6/ABRT

Версия 3proxy последняя отсюда с Github.

Конфиг

`
nserver 127.0.0.1
nscache 65535
nscache6 65535
timeouts 1 5 30 60 180 1800 15 60 15 5
plugin /libexec/AuthPlugin.ld.so AuthPlugin
config /conf/3proxy.cfg
monitor /conf/users
include /conf/users
noforce
auth r_auth strong
allow *
parent 1000 http 11.11.11.11 1200 login password
maxconn 10000
proxy -s0 -osTCP_NODELAY -ocTCP_NODELAY -a -n -i12.12.12.12 -e12.12.12.12 -p1213
flush
fakeresolve
`

Плагин разбивает пароль клиента на 2 части.
Первая часть идет для локальной авторизации через strong
Вторая часть присоединяется к цепочке parent в extpass


Плагин
`
#include "../../structures.h"
#include <string.h>
#include <ctype.h>
#include <stddef.h>
#include <stdlib.h>

#ifdef  __cplusplus
extern "C" {
#endif

static struct auth alwaysauth;
static unsigned char *extpassword;

#ifdef  __cplusplus
extern "C" {
#endif

ptrdiff_t indexof(unsigned char* str, char ch) {
    unsigned char *moved_string = (unsigned char*)strchr((const char*)str, ch);
    return moved_string ? moved_string - str : -1;
}

unsigned char* substring(unsigned char* str, size_t position, size_t length) {
    unsigned char* pointer = malloc(length + 1);
    if(pointer == NULL)
        return NULL;

    int c;
    for(c = 0 ; c < length ; c++) {
        pointer[c] = str[position];
        str++;
    }
    pointer[c] = '\0';

    return pointer;
}

unsigned char* concat(unsigned char *s1, unsigned char *s2) {
    size_t len1 = strlen((const char *)s1);
    size_t len2 = strlen((const char *)s2);

    unsigned char *result = malloc(len1 + len2 + 1);
    if(!result) {
        fprintf(stderr, "malloc() failed: insufficient memory!\n");
        return NULL;
    }

    memcpy(result, (char *)s1, len1);
    memcpy(result + len1, (char *)s2, len2);
    return result;
}

static int aufunc(struct clientparam* param) {
    if(!param->username || !param->password || param->pwtype != 0) {
        return 4;
    }

    struct chain *chain = param->srv && param->srv->acl ? param->srv->acl->chains : NULL;
    if(!chain) {
        return 4;
    }

    int index = indexof(param->password, '_');
    if(index <= -1) {
        return 4;
    }

    unsigned char *pass1 = substring(param->password, 0, index);
    unsigned char *pass2 = substring(param->password, index, strlen((const char *)param->password) - index);

    if(pass1) {
        free(param->password);
        param->password = (unsigned char *)strdup((char *)(pass1));
	    free(pass1);
    }

    if(pass2) {
        if(extpassword)
            free(extpassword);

        index = indexof(chain->extpass, '_');
        if(index > -1)
            pass1 = substring(chain->extpass, 0, index);
        else
            pass1 = (unsigned char *)strdup((char *)chain->extpass);

        extpassword = concat(pass1, pass2);

        free(chain->extpass);
        chain->extpass = (unsigned char *)strdup((char *)(extpassword));
        
        free(pass1);
    	free(pass2);
    }

    return 4;
 }

#ifdef WATCOM
#pragma aux AuthPlugin "*" parm caller [ ] value struct float struct routine [eax] modify [eax ecx edx]
#undef PLUGINCALL
#define PLUGINCALL
#endif

PLUGINAPI int PLUGINCALL AuthPlugin(struct pluginlink * pluginlink, int argc, char** argv) {
    static int loaded = 0;

    if (!loaded) {
        alwaysauth.authenticate = aufunc;
        alwaysauth.authorize = pluginlink->checkACL;
        alwaysauth.desc = "r_auth";
        alwaysauth.next = pluginlink->authfuncs->next;
        pluginlink->authfuncs->next = &alwaysauth;
        loaded = 1;
    }
    return 0;
}


#ifdef  __cplusplus
}
#endif
`

**Comment by srvh** [comment]
Уточню, что в логах нет никаких записей о падении.
Сервис рестартится автоматически через systemd

Плагин самописный на основе плагина utf8tocp1251.
Я представил весь код плагина.

Я копирую плагин в 3proxy/src/plugins/
Добавляю строку в 3proxy/Makefile.Linux в раздел PLUGINS

Затем устанавливаю так:
make -f Makefile
make && make install

В этом случае же основной файл /usr/bin/3proxy обновляется корректно?

**Comment by srvh** [comment]
или нужно удалять /usr/bin/3proxy перед установкой ?

**Comment by z3APA3A** [comment]
Проверь, где ты делаешь двойное освобождение памяти в плагине или освобождаешь память и не зануляешь указатель. Если не делаешь, попробуй использовать не malloc/free/realloc/strdup а то, что передаются в pluginlink, например pluginlink->strdupfunc(), pluginlink->freefunc() и тп

**Comment by srvh** [comment]
Занулить указатель - это присвоить значение NULL ?

Например:
if(pass1) {
        free(param->password);
        param->password = (unsigned char *)strdup((char *)(pass1));
	    free(pass1);

После освобождения памяти для pass1, нужно присвоить NULL, так ?
if(pass1) {
        free(param->password);
        param->password = (unsigned char *)strdup((char *)(pass1));
	    free(pass1);
pass1 = NULL;

**Comment by z3APA3A** [comment]
да, хорошая практика после free присваивать NULL, особенно для тех указателей, которые  используются в нескольких местах
